File: flask_app/app/controllers/subscription_controller.py
from flask import request
from flask_jwt_extended import get_jwt_identity, jwt_required
from app.services.subscription_service import SubscriptionService
from app.utils.response import success_response, error_response

class SubscriptionController:

    @staticmethod
    @jwt_required()
    def get_all():
        page = request.args.get('page', default=1, type=int)
        per_page = request.args.get('per_page', default=10, type=int)

        filters = {
            "user_id": request.args.get("user_id"),
            "is_active": request.args.get("active", type=lambda v: v.lower() == 'true' if v else None),
        }

        filters = {k: v for k, v in filters.items() if v is not None}
    
        subscriptions, total = SubscriptionService.get_all( page, per_page, filters)
        return success_response("Subscriptions retrieved", {
                "items": [s.to_dict() for s in subscriptions],
                "total": total,
                "page": page,
                "per_page": per_page,
            })

    @staticmethod
    @jwt_required()
    def get_one(subscription_id):
        subscription = SubscriptionService.get_by_id(subscription_id)
        if not subscription:
            return error_response("Subscription not found", 404)
        return success_response(subscription.to_dict())

    @staticmethod
    @jwt_required()
    def create():
        user_id = get_jwt_identity()
        data = request.get_json()
        try:
            subscription = SubscriptionService.create(user_id, data)
            return success_response("Subscription Created successfully", subscription.to_dict(), 201)
        except Exception as e:
            return error_response(str(e), 400)

    @staticmethod
    @jwt_required()
    def update(subscription_id):
        data = request.get_json()
        updated = SubscriptionService.update(subscription_id, data)
        if not updated:
            return error_response("Subscription not found", 404)
        return success_response("Subscription Updated successfully",updated.to_dict())

    @staticmethod
    @jwt_required()
    def get_active_subscription(user_id):
        subscription = SubscriptionService.get_active_subscription_by_user(user_id)
        if not subscription:
            return error_response("No active subscription found", 404)
        return success_response("Active subscription found", subscription.to_dict())

    @staticmethod
    @jwt_required()
    def check_auto_renew(user_id):
        subscription = SubscriptionService.get_active_subscription_by_user(user_id)
        if not subscription:
            return error_response("No active subscription found", 404)
        return success_response("Auto-renew status retrieved", {
            "auto_renew": subscription.auto_renew
        })

    @staticmethod
    @jwt_required()
    def delete(subscription_id):
        success = SubscriptionService.soft_delete(subscription_id)
        if not success:
            return error_response("Subscription not found", 404)
        return success_response({"message": "Deleted successfully"})

File: springboot_api/src/main/java/com/example/membership_api/controller/SubscriptionController.java
package com.example.membership_api.controller;

import java.util.HashMap;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.example.membership_api.model.Subscription;
import com.example.membership_api.service.SubscriptionService;
import com.example.membership_api.utils.JwtUtil;
import com.example.membership_api.utils.ResponseUtil;

import jakarta.servlet.http.HttpServletRequest;

@RestController
@RequestMapping("/subscriptions")
public class SubscriptionController {

    @Autowired
    private SubscriptionService subscriptionService;

    @GetMapping
    public ResponseEntity<?> getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int perPage,
            @RequestParam(required = false, name = "user_id") String userId,
            @RequestParam(required = false, name = "active") Boolean active
    ) {
        Map<String, Object> filters = new HashMap<>();
        if (userId != null) filters.put("user_id", userId);
        if (active != null) filters.put("is_active", active);

        Map<String, Object> data = subscriptionService.getAll(page, perPage, filters);
        return ResponseUtil.success("Subscriptions retrieved", data);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getOne(@PathVariable String id) {
        Subscription subscription = subscriptionService.getById(id);
        if (subscription == null) {
            return ResponseUtil.error("Subscription not found", 404);
        }
        return ResponseUtil.success("Subscription found", subscription);
    }

    @PostMapping
    public ResponseEntity<?> create(@RequestBody Map<String, Object> data, HttpServletRequest request) {
        try {
            String userId = JwtUtil.extractUserIdFromRequest(request);
            Subscription created = subscriptionService.create(userId, data);
            return ResponseUtil.success("Subscription created", created);
        } catch (Exception e) {
            return ResponseUtil.error("Failed to create subscription: " + e.getMessage(), 400);
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> update(@PathVariable String id, @RequestBody Map<String, Object> data) {
        Subscription updated = subscriptionService.update(id, data);
        if (updated == null) {
            return ResponseUtil.error("Subscription not found", 404);
        }
        return ResponseUtil.success("Subscription updated", updated);
    }

    @GetMapping("/active/{userId}")
    public ResponseEntity<?> getActiveSubscription(@PathVariable String userId) {
        Subscription subscription = subscriptionService.getActiveSubscriptionByUser(userId);
        if (subscription == null) {
            return ResponseUtil.error("No active subscription found", 404);
        }
        return ResponseUtil.success("Active subscription found", subscription);
    }

    @GetMapping("/auto-renew/{userId}")
    public ResponseEntity<?> checkAutoRenew(@PathVariable String userId) {
        Subscription subscription = subscriptionService.getActiveSubscriptionByUser(userId);
        if (subscription == null) {
            return ResponseUtil.error("No active subscription found", 404);
        }
        return ResponseUtil.success("Auto-renew status", Map.of("auto_renew", subscription.getAutoRenew()));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> delete(@PathVariable String id) {
        boolean deleted = subscriptionService.softDelete(id);
        if (!deleted) {
            return ResponseUtil.error("Subscription not found", 404);
        }
        return ResponseUtil.success("Subscription deleted");
    }
}
